odbc/src/api/odbc_uri.rs

```diff
@@ -0,0 +1,194 @@
+use crate::errors::{ODBCError, Result};
+use std::collections::BTreeMap;
+
+// TODO SQL-990: These errors will probably change.
+const NOT_EMPTY_ERROR: &str = "uri must not be empty";
+const EQUAL_ERROR: &str = "all uri atttributes must be of the form key=value";
+
+// TODO SQL-990: Audit these mandatory attributes
+const USER: &[&str] = &["user", "uid"];
+const PWD: &[&str] = &["pwd", "password"];
+const SERVER: &[&str] = &["server"];
+const SSL: &[&str] = &["ssl"];
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct ODBCUri<'a>(BTreeMap<String, &'a str>);
+
+impl<'a> ODBCUri<'a> {
+    pub fn new(odbc_uri: &'a str) -> Result<ODBCUri<'a>> {
+        if odbc_uri.is_empty() {
+            return Err(ODBCError::InvalidUriFormat(NOT_EMPTY_ERROR.to_string()));
+        }
+        // TODO SQL-990: Support the actual ODBC spec with regards to special characters in attributes
+        // the algorithm will most likely need to be a state machine over each character in the
+        // odbc_uri string.
+        odbc_uri
+            .split(';')
+            .filter(|attr| !attr.is_empty())
+            .map(|attr| {
+                // now split each attribute pair on '='
+                let mut sp = attr.split('=').collect::<Vec<_>>();
+                if sp.len() != 2 {
+                    return Err(ODBCError::InvalidUriFormat(EQUAL_ERROR.to_string()));
+                }
+                // ODBC attribute keys are case insensitive, so we lowercase the keys
+                Ok((sp.remove(0).to_lowercase(), sp.remove(0)))
+            })
+            .collect::<Result<BTreeMap<_, _>>>()
+            .map(ODBCUri)
+    }
+
+    // remove will remove the first value with a given one of the names passed, assuming all names
+    // are synonyms.
+    pub fn remove(&mut self, names: &[&str]) -> Option<&'a str> {
+        for name in names.iter() {
+            let ret = self.0.remove(&name.to_string());
+            if ret.is_some() {
+                return ret;
+            }
+        }
+        None
+    }
+
+    // remove_or_else is the same as remove but with a default thunk.
+    pub fn remove_or_else(&mut self, f: impl Fn() -> &'a str, names: &[&str]) -> &'a str {
+        self.remove(names).unwrap_or_else(f)
+    }
+
+    // remove_mandatory_attribute will find an attribute that must exist and transfer ownership to
+    // the caller. It accepts a slice of names that will be checked in order for names that are
+    // synonyms (e.g., uid and user are both viable attribute names for a user). If both names
+    // exist, it will only find the first.
```

Comment:
This is private because I only foresee its ever being used in `remove_mongo_uri`